### PR TITLE
Feature: pulling consumer subscriber

### PR DESCRIPTION
**Pull request by astelmashenko**
# Proposed Changes

- introducing PullConsumer in parallel with legacy PushConsumer
- switching e2e tests to test jetstream dispatcher

### CODE REVIEW

File: pkg/channel/jetstream/dispatcher/push_consumer.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+	if logger.Desugar().Core().Enabled(zap.DebugLevel) {
+		var debugKVs []interface{}
+		if meta, err := msg.Metadata(); err == nil {
+			debugKVs = append(debugKVs, zap.Any("msg_metadata", meta))
+		}
+
+		logger.Debugw("received message from JetStream consumer", debugKVs...)
+	}
+
+	message := cejs.NewMessage(msg)
```
Comment (Cali0707) [inline pkg/channel/jetstream/dispatcher/push_consumer.go:112]: Shouldn't we call `message.Finish()` somewhere?

File: pkg/channel/jetstream/dispatcher/pull_consumer.go [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+			if !ok {
+				return nil
+			}
+
+			wg.Add(1)
+
+			go func() {
+				defer wg.Done()
+				ctx := logging.WithLogger(ctx, c.logger.With(zap.String("msg_id", natsMsg.Header.Get(nats.MsgIdHdr))))
+				msg := internal.NewMessage(ctx, natsMsg, c.natsConsumerInfo.Config.AckWait)
```
Comment (Cali0707) [inline pkg/channel/jetstream/dispatcher/pull_consumer.go:189]: do we need to call `msg.Finish()` somewhere?

**Review by Cali0707 — COMMENTED** [review]
I'm not very familiar with the NATS-specific stuff happening here, but I did my best to do a first pass review @astelmashenko

File: pkg/channel/jetstream/dispatcher/pull_consumer.go [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+	// configured via the BATCH_GOROUTINES environment variable.
+	//
+	// If you have sufficient resources, it is better to have NumBatchGoRoutines equal to
+	// FetchBatchSize so that messages received by JetStream are immediately processed. This will
+	// give the upstream request the closest amount of time to respond to the consumer's configured
+	// AckWait. This is beneficial because if the eventing subscriber is a scale-to-zero service,
+	// you have more time for resources to spin up.
+	//
+	// If you have limited resources, you may want to reduce the number of goroutines.
+	NumBatchGoRoutines = 32
```
[the remark this answers is not in the document]
Comment (astelmashenko) [reply pkg/channel/jetstream/dispatcher/pull_consumer.go:59]: yeah, you are right, I was planning to introduce later worker pools per subscriber.

File: pkg/channel/jetstream/dispatcher/pull_consumer.go (hunk 2 above)
Comment (astelmashenko) [reply pkg/channel/jetstream/dispatcher/pull_consumer.go:189]: it is called inside message_dispatcher, https://github.com/knative-extensions/eventing-natss/blob/aabb79ea807ba93cd7ca53f469dfa1d459a30969/pkg/channel/jetstream/dispatcher/message_dispatcher.go#L137

File: pkg/channel/jetstream/dispatcher/push_consumer.go (hunk 1 above)
Comment (astelmashenko) [reply pkg/channel/jetstream/dispatcher/push_consumer.go:112]: it is called inside message_dispatcher, https://github.com/knative-extensions/eventing-natss/blob/aabb79ea807ba93cd7ca53f469dfa1d459a30969/pkg/channel/jetstream/dispatcher/message_dispatcher.go#L137

File: pkg/channel/jetstream/dispatcher/pull_consumer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Cali0707) [reply pkg/channel/jetstream/dispatcher/pull_consumer.go:59]: should we remove it for now then?

[the remark this answers is not in the document]
Comment (astelmashenko) [reply pkg/channel/jetstream/dispatcher/pull_consumer.go]: agree. removed.

**Review by Cali0707 — COMMENTED** [review]
I'm not an expert on any of the NATS parts, but this looks fine to me 👍 

/lgtm
/hold for any other reviews

**Comment by astelmashenko** [comment]
@Cali0707 , hey thanks for reviewing it, if you have somebody in mind who can review it then please mention him here, otherwise I do not know why hold, the PR is two months old, enough time to check.
Basically it will not change how it's working by default unless you intentionally pass configuration param to use pulling consumers.

I'm waiting this PR to continue working on the eventing-nats:
- remove dweprecated (and not supported) nats-streaming
- introduce a logic seamlessly to switch from deprecated push-consumers

**Comment by Cali0707** [comment]
Sure @astelmashenko let's merge it then
/unhold

**Comment by astelmashenko** [comment]
Hey, @Cali0707 , could we skip codecov for this PR? In the PR I do:
- refactor code to support push/pull based consumers (push based is default and without changes)
- changed github actions to run tests against jetstream implementation (was for legacy nats-streaming)

Because of refactorings and new code there is such codecov that does not allow to merge.

**Pull request merged**